### DISCUSSION TITLE
Performance Degradation in Routing (rerendering)

### DISCUSSION THREAD

**Discussion by AndrejNemec** [TanStack/router#1460] [General]
I'm reaching out to report a performance issue.

The problem involves performance in two scenarios:

1. When updating the search parameters using `useNavigation`, the entire page, including all layout components and child components, is re-rendered.
2. When there is a layout component with multiple child pages, navigating between these child pages triggers a re-render of the entire page, including all layout components.

In the classic `react-router` package, this problem does not exist, there it works as expected, parts are rerendered as expected.

Example Link: https://stackblitz.com/edit/vitejs-vite-ifxs98

**Comment by rajivharris** [comment]
I am also facing the same issue. The entire page is reloading when any navigation link is clicked.

**Comment by s-hoff** [comment]
I am facing the same, or at least a similar problem. There needs to be a way to "silently" update search params without causing any rerenders or actual navigation.

I want to reflect the state of some of my forms in the search params to make them shareable. The form itself is currently managed by react-hook-form and has the components cut specifically to keep rerenders small. Having the search param updates rerender the whole thing completely defeats the purpose of some of the things I do to avoid it.

The TanStack router docs specifically mention and encourage putting state information into the search params multiple times. In fact, it is what drew me to it. However, this is not viable to do unless the search params can be updated without rerenders.


Thinking about it, the whole Idea of "navigating" to only change the search params to reflect an internal state change seems off - there is no real navigation here, just an additional bit of user interface being updated.

**Reply by s-hoff** [reply]
Upon closer inspection, at least in my case, the Router.useSearch() is whats actually causing the rerender. Is there a way to have that hook only apply when the navigation occurred from a different origin, and not when it is from the same place?
I would prefer to still keep the search params outside of my inputs, which would otherwise be the only way I see to make sure to only rerender what is actually necessary.

**Reply by ViewableGravy** [reply]
You've likely found a solution to this since, but in general, if you want any of the hooks in tanstack router to avoid re-rendering unecessarily, you'll want to use the `select` function to select just what you need. From there, those hooks will only re-render your component when the result of the select function changes.

For example, if you need to access a search param such as `sort_by`, you'd do something like follows
```ts
const sort_by = useSearch({
  from: "/my-route",
  select: ({ sort_by }) => sort_by
})
```

I imagine it would be possible to prevent these re-renders within the router (it may already do it with `structuralSharing` enabled, although I haven't tested this), but at least in my case where we have been on version `1.51.0` for a while, and recently upped it to `1.86.0`, using a select has been the only way to prevent any route related changes causing a re-render from hooks such as `useSearch` or `useLocation`

**Reply by s-hoff** [reply]
My apologies for not adding it here, but yes, I got the solution on the discord later.
I misunderstood what `useSearch` is for in relation to my use case: It _subscribes_ to the search params, while I was only interested in a one-time snapshot when (re)loading the page and wanted to store the forms state in the search params for sharing and reloading only.
The solution was to grab it from `router.state.location.search` instead:
https://stackblitz.com/edit/tanstack-router-yfaf7nps?file=src%2Froutes%2Findex.tsx&preset=node

I suppose if one additionally wants to subscribe to some of the search params, grabbing those with a select function would then make both work.

I'm not sure how well this applies to the others problems.
